### PR TITLE
Ana/fix gql queries not coming from cache

### LINKED ISSUE #4367: GraphQL requests not comming from cache
**Is your feature request related to a problem? Please describe.**

When opening the ActionModal we see, that requests are done against the API which slow down opening the ActionModal.

**Describe the solution you'd like**

The requests should be cached already and therefore should be instant. We need to investigate why this is not the case.
In the case of the ActionModal we should also investigate which requests are actually needed to be done there. The validators in Polkadot are 500 which is a huge payload. Maybe we can reduce them by defaulting to popular.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #4367 

**Description:**

One line change 😛 

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by Bitcoinera** [comment]
This is what it takes now:

**Comment by faboweb** [comment]
🚀 Deployed on https://5ef5172d579e3a94f344e564--lunieio.netlify.app

File: app/src/gql/apollo.js [hunk 1]
```diff
@@ -134,7 +134,7 @@ export const createApolloProvider = async () => {
     defaultOptions: {
       // apollo options applied to all queries in components
       $query: {
-        fetchPolicy: "cache-and-network",
+        fetchPolicy: "cache-first",
```
Comment (faboweb) [inline app/src/gql/apollo.js:137]: cache-and-network: This fetch policy will have Apollo first trying to read data from your cache. If all the data needed to fulfill your query is in the cache then that data will be returned. However, regardless of whether or not the full data is in your cache this fetchPolicy will always execute query with the network interface unlike cache-first which will only execute your query if the query data is not in your cache. This fetch policy optimizes for users getting a quick response while also trying to keep cached data consistent with your server data at the cost of extra network requests.

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: This is the theory, but now try it

Comment (faboweb) [reply app/src/gql/apollo.js:137]: Problem with cache first is, that it will not update. So why does cache-and-network not use existing queries?

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: Ok, will investigate

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: The problem seems to be the `loading` prop. We should remove it if we want the results to be coming from cache: https://github.com/apollographql/react-apollo/issues/2739

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: We can stop relying on loading and only look at `this.balances.length` and so on... I am going to give it a try

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: Nothing... 😞 I was also reading some SmartQueries concrete threads but cannot make it work (get results from cache) other than applying `cache-first`.

Look at this article: https://medium.com/@galen.corey/understanding-apollo-fetch-policies-705b5ad71980

I think if we explicitly have the `updates` (like we in fact do) in our SmartQueries we should be fine actually

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: I can open a Draft PR with the "no-apollo-loading" rewrite but I tested it and only `cache-fist` seems to be really working

Comment (Bitcoinera) [reply app/src/gql/apollo.js:137]: OK, this is not a good solution. I tried undelegating and had to do a hard refresh in order to be able to see the undelegation in PagePortfolio... So not ideal

**Comment by Bitcoinera** [comment]
OK, after reading [here ](https://www.apollographql.com/docs/react/caching/cache-interaction/) about the `cacheRedirects `and [here ](https://www.apollographql.com/docs/react/caching/cache-configuration/#custom-identifiers) about custom identifiers, I think I finally found the root of our problem with GQL queries: 
we should be using "id" fields for all of our GQL entities, same that we do with `Network` and `Proposal`.

Otherwise, the query response will fall in cache under the `ROOT_QUERY`, which is not good.

This can be visually confirmed with a breakpoint in the connection module and then entering `apollo.cache.data.data` (I set the breakpoint on the `const { data } = apollo.query` line).

Another option is to set custom identifiers like I was trying to do in the `add custom identifiers for queries`, but I think it is more elegant to use the `id` solution.

I noticed an improvement of performance by adding the identifiers and cacheRedirects (these last one I am still unsure if we need them or not)

Should I go then for the `ids` solution?

**Comment by faboweb** [comment]
Yes, let'S add IDs for all the objects.

**Comment by Bitcoinera** [comment]
Mmm, tried adding ids to rewards and delegations and was checking it in `PageValidator`, but performance seemed to decrease instead of increase :woman_shrugging: It was taking a while to load it. But now with ids in the validators query I can see a great improvement. Even in ActionModal for some reason

**Comment by Bitcoinera** [comment]
Need to keep testing I think

File: api/lib/notifications.js [hunk 2]
```diff
@@ -220,6 +220,7 @@ const startNotificationService = () => {
       const notificationResponse =
         response.data.insert_notifications.returning[0]
       const notification = {
+        id: getMessageTitle(notificationResponse),
```
Comment (faboweb) [inline api/lib/notifications.js:223]: a title is not a good id. get the actual id of the notification from the db?

File: api/lib/reducers/cosmosV0-reducers.js [hunk 3]
```diff
@@ -233,6 +234,7 @@ function validatorReducer(networkId, signedBlocksWindow, validator) {
 
 function blockReducer(networkId, block, transactions, data = {}) {
   return {
+    id: block.block_meta.header.height,
```
Comment (faboweb) [inline api/lib/reducers/cosmosV0-reducers.js:237]: that would conflict between different chains, right?
```suggestion
    id: `${networkId}_${block.block_meta.header.height}`,
```

File: api/lib/reducers/kavaV1-reducers.js [hunk 4]
```diff
@@ -6,6 +6,7 @@ function setTransactionSuccess(transaction) {
 
 function blockReducer(networkId, block, transactions, data = {}) {
   return {
+    id: block.block.header.height,
```
Comment (faboweb) [inline api/lib/reducers/kavaV1-reducers.js:9]: see above

File: api/lib/reducers/livepeerV0-reducers.js [hunk 5]
```diff
@@ -3,6 +3,7 @@ const LPT_CONVERSION = `1000000000000000000`
```
Comment (faboweb) [inline api/lib/reducers/livepeerV0-reducers.js:3]: let's remove livepeer files

File: api/lib/reducers/polkadotV0-reducers.js [hunk 6]
```diff
@@ -16,6 +16,7 @@ function blockReducer(
   data = {}
 ) {
   return {
+    id: blockHeight,
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:19]: see above

File: api/lib/reducers/cosmosV0-reducers.js (hunk 3 above)
Comment (faboweb) [reply api/lib/reducers/cosmosV0-reducers.js:237]: or use block hash?

File: api/lib/reducers/cosmosV0-reducers.js [hunk 7]
```diff
@@ -407,6 +414,7 @@ async function reduceFormattedRewards(
         ? await calculateFiatValue(lunieCoin, fiatCurrency)
         : undefined
       multiDenomRewardsArray.push({
+        id: lunieCoin.denom,
```
Comment (faboweb) [inline api/lib/reducers/cosmosV0-reducers.js]: do we need to specify an extra field `id` or can we specify in GQL what the ID field is?

File: api/lib/reducers/cosmosV0-reducers.js (hunk 3 above)
Comment (Bitcoinera) [reply api/lib/reducers/cosmosV0-reducers.js:237]: :thinking: maybe better yes

File: api/lib/reducers/kavaV1-reducers.js (hunk 4 above)
Comment (Bitcoinera) [reply api/lib/reducers/kavaV1-reducers.js:9]: ok, let me change that

File: api/lib/reducers/cosmosV0-reducers.js (hunk 7 above)
Comment (Bitcoinera) [reply api/lib/reducers/cosmosV0-reducers.js]: you mean when we do the query? Not sure... you think ID is a special field? Let's google :sweat_smile:

File: api/lib/notifications.js (hunk 2 above)
Comment (Bitcoinera) [reply api/lib/notifications.js:223]: changed

**Comment by faboweb** [comment]
Looking good! Whats missing?

**Comment by Bitcoinera** [comment]
> Looking good! Whats missing?

Missing time xD But today I'm killing this one off :+1:

**Comment by Bitcoinera** [comment]
Mmm, getting kind of stuck with this error:

image

Have tried couple of things already but what I don't really get is why it is trying to overwrite a validator Object with a delegation one in cache...

**Comment by Bitcoinera** [comment]
Working well. I found this other error just now and tried to fix it with the `map` method in `PageNotifications`. Unfortunately, hasn't worked. But it can be another PR, as it is unrelated to this one:

image

**Comment by Bitcoinera** [comment]
Ready to go! :woman_dancing:

File: .vscode/launch.json [hunk 8]
```diff
@@ -14,8 +14,8 @@
             "program": "${workspaceFolder}/api/index.js",
             "envFile": "${workspaceFolder}/api/.env",
             "env": {
-                "WHITELIST_ORIGIN": "http://localhost:4000",
-                // "RUN_ONLY_NETWORK": "cosmos-hub-testnet"
```
Comment (jbibla) [inline .vscode/launch.json]: should this be in the .vscode file?

Comment (Bitcoinera) [reply .vscode/launch.json]: OK, I just deleted this whole `env` block from the `launch.json`

**Comment by faboweb** [comment]
e2e fails

**Comment by Bitcoinera** [comment]
Mmm, where are the validators for the E2E to be found?

**Comment by faboweb** [comment]
Still failing e2e tests

**Comment by Bitcoinera** [comment]
> Still failing e2e tests

Yes, because of this:

image


We would need to set the E2E to run against the changed API from this PR. Is there no way to do that? :thinking: GitHub Instances, I don't know. We can be noisy with the GitHub people

**Comment by faboweb** [comment]
Need to split the API from the PR and deploy independently

**Comment by Bitcoinera** [comment]
> Need to split the API from the PR and deploy independently

:scream:

**Comment by Bitcoinera** [comment]
Or we can E2E manually, which I already did, and see that it is effectively working, which it is :wink: And we save up dev time

**Comment by Bitcoinera** [comment]
Or we can E2E manually, which I already did, and see that it is effectively working, which it is :wink: And we save up dev time

**Pull request merged**